experiments/middleware.py

```diff
@@ -1,13 +1,33 @@
+from django.conf import settings
+from importlib import import_module
+
 from experiments.utils import participant
 
+try:
+    # for Django >= 1.10
+    from django.utils.deprecation import MiddlewareMixin
+except ImportError:
+    # for Django < 1.10
+    MiddlewareMixin = object
+
+class ExperimentsRetentionMiddleware(MiddlewareMixin):
+
+    # should we create session in process_request ?
 
-class ExperimentsRetentionMiddleware(object):
     def process_response(self, request, response):
         # Don't track, failed pages, ajax requests, logged out users or widget impressions.
         # We detect widgets by relying on the fact that they are flagged as being embedable
         if response.status_code != 200 or request.is_ajax() or getattr(response, 'xframe_options_exempt', False):
             return response
 
+        # create session if not exists     see http://stackoverflow.com/a/5131421/127114
+        engine = import_module(settings.SESSION_ENGINE)
```

Comment:
You shouldn't have to worry about Googlebot creating lots of sessions - the optimisation belongs at a different layer, not here.

Considering that cookies are required to store the SESSION_ID, and that clearing the browser cache will clear said SESSION_ID, using [cookie-based sessions](https://docs.djangoproject.com/en/1.11/topics/http/sessions/#using-cookie-based-sessions) isn't such a bad idea.  You don't have to worry about storage, you reduce latency (not having a round trip to the DB/Cache to retrieve session data), and the "stale" sessions clear themselves.